[Diagnostics] Diagnose subscript operator misuse via fixes

xedin opened the pull request:
Fix to use subscript operator instead of spelled out name helps
to produce a solution, that makes it much easier to diagnose
problems precisely and provide proper fix-its, it also helps to
diagnose ambiguous cases, and stacks up nicely with other errors.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSFix] Attempt to replace `.subscript(...)` with subscript operator

If there are no members named "subscript" available, let's try to
replace it with subscript operator with might be what was intended.

xedin committed:
[Diagnostics] Diagnose subscript operator misuse via fixes

Fix to use subscript operator instead of spelled out name helps
to produce a solution, that makes it much easier to diagnose
problems precisely and provide proper fix-its, it also helps to
diagnose ambiguous cases, and stacks up nicely with other errors.

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -64,6 +64,9 @@ ERROR(ambiguous_reference_to_decl,none,
       "ambiguous subscript with base type %0 and index type %1",
       (Type, Type))
+NOTE(ambiguous_subscript_candidate_requires_operator,none,
+     "candidate requires use of the subscript operator", ())
```
@jrose-apple I'm not really happy with this message, could you please help me to phrase this better? The problem is that once name "subscript" is turned into subscript operator it could be ambiguous like in one of the examples (e.g. based on result type), but I don't only want to point out ambiguity but also mention that fixing it would require using subscript operator as well...

jrose-apple replied:
I feel like the error message itself is misleading, and the note doesn't cover enough. But in cases where you're going to have a note, maybe this can replace "found this candidate" instead of being emitted in addition to it?

(The reason this matters at all is because ```func `subscript`(_ sender: Any?)``` exists.)

xedin replied:
Unfortunately we aren't there yet to replace default "ambiguous member" diagnostic for multiple solutions with something better and add fix-its, so I can only provide additional information with notes and specially I only use this note if the candidate we found has to use subscript operator fix e.g.

```swift
struct S {
  subscript(_ x: String) -> Int { return 42 }
  subscript(_ x: String) -> String { return "42" }
}

func foo(_ s: S) {
  _ = s.subscript("ultimate question")
}
```

Is going to produce error like this now:

```
error: ambiguous reference to subscript 'subscript(_:)'
    _ = s.subscript("ultimate question")
          ^
note: candidate requires use of the subscript operator
    subscript(_ x: String) -> Int { return 42 }
    ^
note: candidate requires use of the subscript operator
    subscript(_ x: String) -> String { return "42" }
    ^
```

> (The reason this matters at all is because func `subscript`(_ sender: Any?) exists.)

Sorry I don't understand what you mean completely... When such overload exists fix that I added would not be attempted.

jrose-apple replied:
> Sorry I don't understand what you mean completely... When such overload exists fix that I added would not be attempted.

We probably already have tests for this, so I won't worry about it.

I'll conclude that I don't think it's worth doing this work until we can change the other diagnostics, then. I find the new top-level errors *less* clear than the old ones.

xedin replied:
The reason why I'm doing this now is because I'm trying to get missing members to be diagnosed via fixes, this is one of the things which came up just like `member used as a property` error. I think that reporting ambiguity here is an improvement as well as couple of other errors related to incorrect arguments where it used to just diagnose as `use subscript operator`, because once I make it so re-labeling is diagnosed via fixes we'd get a nice multi-error suggestion with fix-its instead of current chain of errors.

xedin replied:
How do you determine which is the main problem? Because if I replace `subscript` with subscript operator it would result in ambiguous error, or if I fix ambiguity by adding contextual type it would then result in error about subscript misuse...

It would  probably be best to attach fix-it to each note, but I don't believe we can do that at the moment. At least new ambiguous error combines information about both problems - ambiguity + need to use subscript operator.

xedin replied:
I don't think this change contradicts what you are saying, original behavior is going to be preserved - your example is still going to diagnose as `cannot convert value of type 'String' to expected argument type 'Int'`. The only way to produce ambiguous diagnostic - if there were no overloads for `subscript` name and overloads of the subscript operator produce solutions which we can't disambiguate.

jrose-apple replied:
Right. So, given the existing behavior when there *are* methods named 'subscript', the spelling of the call should be considered more important than the types of the arguments matching up.

xedin replied:
Wait, I'm confused - in previous comment you mentioned that in you'd expect example to produce argument related error (which is what it currently does), but now you are saying that it should actually produce suggestion to replace name with operator?

jrose-apple replied:
I'm sorry if I've been unclear. I've been trying to state that "you wrote `subscript` when you should have used brackets" is the error you should be told about first.

xedin replied:
I understood that, but it contradicts your previous comment given its example, also current behavior is consistent with it and is unchanged by this fix.

xedin replied:
Ok, I think I understand what you mean now so the actual diagnostic then would be something like: 

```swift
error: value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?

note: found this candidate:
    subscript(_ x: String) -> Int { return 42 }
    ^
note: found this candidate:
    subscript(_ x: String) -> String { return "42" }
    ^
```

xedin replied:
Done!

xedin commented:
@swift-ci please smoke test

xedin committed:
[Diagnostics] Diagnose ambiguity resulting from subscript operator fix as a missing member

Since the rule is to prioritize names over types, let's diagnose
ambiguous solutions containing subscript operator fix as missing
member and list possible candidates to use.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

jrose-apple commented on test/decl/subscript/subscripting.swift:
```diff
@@ -285,21 +287,23 @@ func testSubscript1(_ s1 : SubscriptTest1) {
+  // expected-note@-2 {{overloads for 'subscript' exist with these partially matching parameter lists: (Protocol), (String), (SubClass)}}
   _ = s1.subscript("hello")
-  // expected-error@-1 {{value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?}} {{9-10=}} {{10-19=}} {{19-20=[}} {{27-28=]}}
+  // expected-error@-1 {{value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?}}
```
Did we *lose* the fix-its here?

jrose-apple commented on test/decl/subscript/subscripting.swift:
```diff
@@ -285,21 +287,23 @@ func testSubscript1(_ s1 : SubscriptTest1) {
   // expected-error@-1 {{value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?}} {{9-10=}} {{10-19=}} {{19-20=[}} {{54-55=]}}
   _ = s1.subscript(true)
-  // expected-error@-1 {{value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?}}
+  // expected-error@-1 {{cannot invoke 'subscript' with an argument list of type '(Bool)'}}
```
I guess you disagree with me on this?

xedin replied:
No, I don't really disagree, I'm just going to make it work with re-labeling/re-typing fix. It's going to have both diagnostics for incorrect argument to fix-it about `subscript`. It's on my radar, I just need to gradually fix CSDiag mess, because that's what is the biggest barrier here.

jrose-apple replied:
Okay, I'll stop bugging you about this. :-) I definitely think you've been doing great chipping away at CSDiag!

xedin replied:
It's okay, no problem! :) Once I convert most of the stuff to use fixes we wouldn't have this ordering problem anymore, but for a bit some of the diagnostics might be in reverse order to what they used to be before, because it would fall into CSDiag first and then diagnose with a fix...
